File: packages/protocol/src/schemas/events.ts
/**
 * Bridge → phone messages — v11 mirror of `crates/protocol/src/events.rs`,
 * which is authoritative.
 *
 * Storage class per message (see kinds.ts):
 * - `sessions` rides SESSION_LIST_KIND (replaceable heartbeat).
 * - `output`, `usage`, `gsd-state` ride LIVE_KIND (ephemeral — loss recoverable
 *   via sync / re-request).
 * - Everything else (sync chunks, acks, lifecycle, pairing) rides RESPONSE_KIND
 *   (stored, 1h expiry) so a briefly-offline phone still receives it.
 */
import { z } from 'zod';
import {
  agentDescriptorSchema,
  credentialStatusSchema,
  gsdStateSchema,
  outputEntrySchema,
  providerProfileInfoSchema,
  remoteSessionInfoSchema,
  sessionOptionSchema,
  usageDataSchema,
} from './common';
import { seqRangeSchema } from './commands';

// --- Session list heartbeat ---

export const sessionListMessageSchema = z.object({
  type: z.literal('sessions'),
  machine: z.string().min(1),
  /** Which host binary publishes this list (UI badge; identity is the keypair). */
  host: z.enum(['cli', 'vscode', 'service']).optional(),
  sessions: z.array(remoteSessionInfoSchema),
  /** The agent backends this bridge can run sessions on. */
  agents: z.array(agentDescriptorSchema),
  /** The bridge's own credentials (not tied to an agent), e.g. a GitHub token. */
  credentials: z.array(credentialStatusSchema).default([]),
  protocolVersion: z.number().int().positive(),
  capabilities: z.array(z.string()).optional(),
  /** Project folders per workspace root, relative paths. Valid `create-session.cwd` values. */
  folders: z.array(z.string()).optional(),
  /**
   * CDX-031: the workspace roots themselves, ABSOLUTE, in `--workspace` order.
   *
   * `folders` lists what is *inside* the roots and cannot name a root: a root
   * never appears among its own children, and two roots' children are unioned
   * into one flat list with no root identity. So with `--workspace a
   * --workspace b` the phone had no way to say "start in b" — the picker
   * offered `a`'s and `b`'s subfolders and every relative path resolves
   * against the FIRST root that contains it.
   *
   * Absolute, addressed exactly like `create-folder.root`, and a valid
   * `create-session.cwd`: `resolveSessionCwdMulti` matches an absolute request
   * against each root's containment check, so root N resolves to itself and a
   * path outside every root still falls back to the first root.
   */
  roots: z.array(z.string()).optional(),
  /** v10: explicit tombstones. The ONLY way a bridge removes a session from the
   *  phone — absence from `sessions` alone never deletes (the old behaviour
   *  silently wiped phones whenever a bridge published a short list). */
  removedSessions: z.array(z.string()).optional(),
  /** v10: set on clean shutdown — sessions remain listed (state: 'offline'). */
  machineOffline: z.boolean().optional(),
});

// --- Live output ---

export const outputMessageSchema = z.object({
  type: z.literal('output'),
  sessionId: z.string().min(1),
  seq: z.number().int().nonnegative(),
  entry: outputEntrySchema,
});

/** v10: confirms receipt of an `input` command (echoes `inputId`) — the phone's
 *  outbox flips pending → confirmed on this, or surfaces failure on timeout. */
export const inputAckMessageSchema = z.object({
  type: z.literal('input-ack'),
  sessionId: z.string().min(1),
  inputId: z.string().min(1),
});

// --- Transcript sync (v10) ---

export const syncBeginMessageSchema = z.object({
  type: z.literal('sync-begin'),
  sessionId: z.string().min(1),
  syncId: z.string().min(1),
  /** Highest persisted seq at sync start — the phone's completeness target. */
  seqHigh: z.number().int().nonnegative(),
  /** Ranges this sync will deliver (the complement of the request's haveRanges). */
  ranges: z.array(seqRangeSchema),
});

export const syncChunkMessageSchema = z.object({
  type: z.literal('sync-chunk'),
  sessionId: z.string().min(1),
  syncId: z.string().min(1),
  /** Inclusive seq range covered by this chunk. */
  range: seqRangeSchema,
  entries: z.array(
    z.object({
      seq: z.number().int().nonnegative(),
      entry: outputEntrySchema,
    }),
  ),
});

export const syncEndMessageSchema = z.object({
  type: z.literal('sync-end'),
  sessionId: z.string().min(1),
  syncId: z.string().min(1),
  /** What was actually delivered (after retries) — the phone re-requests the
   *  difference on next connect instead of assuming completeness. */
  deliveredRanges: z.array(seqRangeSchema),
});

// --- Two-phase session creation ---

export const sessionPendingMessageSchema = z.object({
  type: z.literal('session-pending'),
  pendingId: z.string().min(1),
  machine: z.string(),
  createdAt: z.string(),
});

export const sessionReadyMessageSchema = z.object({
  type: z.literal('session-ready'),
  pendingId: z.string().min(1),
  session: remoteSessionInfoSchema,
});

export const sessionFailedMessageSchema = z.object({
  type: z.literal('session-failed'),
  pendingId: z.string().min(1),
  reason: z.string(),
});

// --- Lifecycle acks / notices ---

export const inputFailedMessageSchema = z.object({
  type: z.literal('input-failed'),
  sessionId: z.string().min(1),
  /** 'no-session': the bridge knows no such session; 'error': the session
   *  exists but rejected the input (dead/ended runner); 'busy': reserved for
   *  a runner that cannot take input right now; 'expired': the command aged
   *  out before the bridge saw it. */
  reason: z.enum(['no-session', 'expired', 'busy', 'error']),
  /** Echoed when the failing input carried an id (outbox correlation). */
  inputId: z.string().optional(),
});

export const closeSessionAckMessageSchema = z.object({
  type: z.literal('close-session-ack'),
  sessionId: z.string().min(1),
  success: z.boolean(),
});

export const sessionReplacedMessageSchema = z.object({
  type: z.literal('session-replaced'),
  oldSessionId: z.string().min(1),
  newSession: remoteSessionInfoSchema,
});

/** A session option now has `value` — the reply to `set-option`, and also
 *  sent when the agent changes an option on its own (e.g. entering plan mode). */
export const optionConfirmedMessageSchema = z.object({
  type: z.literal('option-confirmed'),
  sessionId: z.string().min(1),
  option: sessionOptionSchema,
  value: z.string(),
});

export const folderAckMessageSchema = z.object({
  type: z.literal('folder-ack'),
  requestId: z.string().min(1),
  success: z.boolean(),
  /** Created folder path relative to its root (usable as create-session.cwd). */
  path: z.string().optional(),
  error: z.string().optional(),
});

// --- Status snapshots ---

export const usageMessageSchema = z.object({
  type: z.literal('usage'),
  sessionId: z.string().min(1),
  usage: usageDataSchema,
});

export const gsdStateMessageSchema = z.object({
  type: z.literal('gsd-state'),
  sessionId: z.string().min(1),
  gsd: gsdStateSchema,
});

/** An agent's live model list. */
export const modelsMessageSchema = z.object({
  type: z.literal('models'),
  /** Echoes the request's `agent`. */
  agent: z.string().min(1),
  models: z.array(
    z.object({
      id: z.string().min(1),
      label: z.string().optional(),
    }),
  ),
  defaultModel: z.string().optional(),
  /** CDX-035: why the bridge could not answer with a list. Set ONLY alongside
   *  an empty `models` — an empty list is "could not answer", never "this
   *  agent supports zero models", so the phone shows the reason, keeps
   *  whatever list it already had, and keeps re-requesting. */
  error: z.string().optional(),
});

// --- Credentials / device config / pairing ---

/** Reply to `set-credentials`: the resulting status of every credential in
 *  the written scope (`agent`, or the bridge's own when absent). */
export const credentialsAckMessageSchema = z.object({
  type: z.literal('credentials-ack'),
  machine: z.string(),
  agent: z.string().optional(),
  success: z.boolean(),
  credentials: z.array(credentialStatusSchema),
  error: z.string().optional(),
});

export const deviceConfigAckMessageSchema = z.object({
  type: z.literal('device-config-ack'),
  success: z.boolean(),
  reachable: z.boolean().optional(),
  error: z.string().optional(),
});

export const pairAckMessageSchema = z.object({
  type: z.literal('pair-ack'),
  machine: z.string(),
  ok: z.boolean(),
  reason: z.enum(['bad-token', 'window-closed']).optional(),
  /** The bridge's relay list, so a manual-npub pairing (which carries no
   *  relays in its URL) still learns where this bridge actually lives. The
   *  phone merges these into its settings (deduped). */
  relays: z.array(z.string()).optional(),
  /** Which host binary acked — same badge vocabulary as the heartbeat. */
  host: z.enum(['cli', 'vscode', 'service']).optional(),
});

// --- Custom AI provider profiles (CDX-062) ---

/** The bridge's redacted stored-profile list (each entry carries `hasToken`,
 *  never the token itself). Rides RESPONSE_KIND (stored) so a briefly-offline
 *  phone still receives it. Deliberately NO `error` field — unlike `models`
 *  (which needs a live SDK to answer), this is always answerable straight from
 *  bridge storage, so there is no "could not answer" case to report. */
export const providerProfilesMessageSchema = z.object({
  type: z.literal('provider-profiles'),
  machine: z.string(),
  profiles: z.array(providerProfileInfoSchema),
});

/** Ack for one `set-provider-profile`. Rides RESPONSE_KIND (stored).
 *  `tokenValid` is tri-state like credentials-ack's `keyValid`: true/false =
 *  live probe verdict, absent = the probe could not run (network error). */
export const providerProfileAckMessageSchema = z.object({
  type: z.literal('provider-profile-ack'),
  machine: z.string(),
  profileId: z.string().min(1),
  success: z.boolean(),
  tokenValid: z.boolean().optional(),
  error: z.string().optional(),
});

// --- Union ---

export const bridgeToPhoneSchema = z.union([
  sessionListMessageSchema,
  outputMessageSchema,
  inputAckMessageSchema,
  syncBeginMessageSchema,
  syncChunkMessageSchema,
  syncEndMessageSchema,
  sessionPendingMessageSchema,
  sessionReadyMessageSchema,
  sessionFailedMessageSchema,
  inputFailedMessageSchema,
  closeSessionAckMessageSchema,
  sessionReplacedMessageSchema,
  optionConfirmedMessageSchema,
  folderAckMessageSchema,
  usageMessageSchema,
  gsdStateMessageSchema,
  modelsMessageSchema,
  credentialsAckMessageSchema,
  deviceConfigAckMessageSchema,
  pairAckMessageSchema,
  providerProfilesMessageSchema,
  providerProfileAckMessageSchema,
]);

export type SessionListMessage = z.infer<typeof sessionListMessageSchema>;
export type OutputMessage = z.infer<typeof outputMessageSchema>;
export type InputAckMessage = z.infer<typeof inputAckMessageSchema>;
export type SyncBeginMessage = z.infer<typeof syncBeginMessageSchema>;
export type SyncChunkMessage = z.infer<typeof syncChunkMessageSchema>;
export type SyncEndMessage = z.infer<typeof syncEndMessageSchema>;
export type SessionPendingMessage = z.infer<typeof sessionPendingMessageSchema>;
export type SessionReadyMessage = z.infer<typeof sessionReadyMessageSchema>;
export type SessionFailedMessage = z.infer<typeof sessionFailedMessageSchema>;
export type InputFailedMessage = z.infer<typeof inputFailedMessageSchema>;
export type CloseSessionAckMessage = z.infer<typeof closeSessionAckMessageSchema>;
export type SessionReplacedMessage = z.infer<typeof sessionReplacedMessageSchema>;
export type OptionConfirmedMessage = z.infer<typeof optionConfirmedMessageSchema>;
export type FolderAckMessage = z.infer<typeof folderAckMessageSchema>;
export type UsageMessage = z.infer<typeof usageMessageSchema>;
export type GsdStateMessage = z.infer<typeof gsdStateMessageSchema>;
export type ModelsMessage = z.infer<typeof modelsMessageSchema>;
export type CredentialsAckMessage = z.infer<typeof credentialsAckMessageSchema>;
export type DeviceConfigAckMessage = z.infer<typeof deviceConfigAckMessageSchema>;
export type PairAckMessage = z.infer<typeof pairAckMessageSchema>;
export type ProviderProfilesMessage = z.infer<typeof providerProfilesMessageSchema>;
export type ProviderProfileAckMessage = z.infer<typeof providerProfileAckMessageSchema>;
export type BridgeToPhoneMessage = z.infer<typeof bridgeToPhoneSchema>;

File: crates/client-core/src/wire/events.rs
//! Bridge → phone messages. Port of `packages/protocol/src/schemas/events.ts`.
//! Storage class per message is in `kinds.rs`. The `BridgeToPhone` union is
//! `#[serde(tag = "type")]`.

use serde::{Deserialize, Serialize};

use super::capabilities::BridgeHostKind;
use super::common::{
    AuthStatus, EffortLevel, GsdState, OutputEntry, PermissionMode, ProviderProfileInfo,
    RemoteSessionInfo, UsageData,
};
use crate::ranges::SeqRange;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionListMsg {
    pub machine: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub host: Option<BridgeHostKind>,
    pub sessions: Vec<RemoteSessionInfo>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth_status: Option<AuthStatus>,
    pub protocol_version: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub capabilities: Option<Vec<String>>,
    /// project folders per workspace root (relative). Valid `create-session.cwd`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub folders: Option<Vec<String>>,
    /// CDX-031: the workspace roots themselves, ABSOLUTE, in `--workspace` order.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub roots: Option<Vec<String>>,
    /// v10: explicit tombstones — the ONLY way a bridge removes a session
    /// (absence from `sessions` never deletes).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub removed_sessions: Option<Vec<String>>,
    /// v10: set on clean shutdown — sessions stay listed (`state: offline`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub machine_offline: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputMsg {
    pub session_id: String,
    pub seq: u64,
    pub entry: OutputEntry,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InputAckMsg {
    pub session_id: String,
    pub input_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncBeginMsg {
    pub session_id: String,
    pub sync_id: String,
    pub seq_high: u64,
    pub ranges: Vec<SeqRange>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncEntry {
    pub seq: u64,
    pub entry: OutputEntry,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncChunkMsg {
    pub session_id: String,
    pub sync_id: String,
    pub range: SeqRange,
    pub entries: Vec<SyncEntry>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncEndMsg {
    pub session_id: String,
    pub sync_id: String,
    pub delivered_ranges: Vec<SeqRange>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionPendingMsg {
    pub pending_id: String,
    pub machine: String,
    pub created_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionReadyMsg {
    pub pending_id: String,
    pub session: RemoteSessionInfo,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionFailedMsg {
    pub pending_id: String,
    pub reason: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum InputFailedReason {
    NoSession,
    Expired,
    Busy,
    Error,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InputFailedMsg {
    pub session_id: String,
    pub reason: InputFailedReason,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub input_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CloseSessionAckMsg {
    pub session_id: String,
    pub success: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionReplacedMsg {
    pub old_session_id: String,
    pub new_session: RemoteSessionInfo,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModeConfirmedMsg {
    pub session_id: String,
    pub mode: PermissionMode,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EffortConfirmedMsg {
    pub session_id: String,
    pub level: EffortLevel,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelConfirmedMsg {
    pub session_id: String,
    pub model: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderAckMsg {
    pub request_id: String,
    pub success: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UsageMsg {
    pub session_id: String,
    pub usage: UsageData,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GsdStateMsg {
    pub session_id: String,
    pub gsd: GsdState,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ModelEntry {
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelsMsg {
    pub models: Vec<ModelEntry>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub default_model: Option<String>,
    /// CDX-035: why the bridge could not answer — set ONLY alongside an empty
    /// `models`; the phone keeps its list and keeps re-requesting.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CredentialsAckMsg {
    pub machine: String,
    pub success: bool,
    pub has_anthropic_key: bool,
    pub has_github_pat: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub key_valid: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeviceConfigAckMsg {
    pub success: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reachable: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PairAckReason {
    BadToken,
    WindowClosed,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PairAckMsg {
    pub machine: String,
    pub ok: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<PairAckReason>,
    /// the bridge's relay list, so a manual-npub pairing still learns where it
    /// lives (merged into settings, deduped).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub relays: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub host: Option<BridgeHostKind>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderProfilesMsg {
    pub machine: String,
    pub profiles: Vec<ProviderProfileInfo>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderProfileAckMsg {
    pub machine: String,
    pub profile_id: String,
    pub success: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub token_valid: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// The bridge→phone message union.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum BridgeToPhone {
    Sessions(SessionListMsg),
    Output(OutputMsg),
    InputAck(InputAckMsg),
    SyncBegin(SyncBeginMsg),
    SyncChunk(SyncChunkMsg),
    SyncEnd(SyncEndMsg),
    SessionPending(SessionPendingMsg),
    SessionReady(SessionReadyMsg),
    SessionFailed(SessionFailedMsg),
    InputFailed(InputFailedMsg),
    CloseSessionAck(CloseSessionAckMsg),
    SessionReplaced(SessionReplacedMsg),
    ModeConfirmed(ModeConfirmedMsg),
    EffortConfirmed(EffortConfirmedMsg),
    ModelConfirmed(ModelConfirmedMsg),
    FolderAck(FolderAckMsg),
    Usage(UsageMsg),
    GsdState(GsdStateMsg),
    Models(ModelsMsg),
    CredentialsAck(CredentialsAckMsg),
    DeviceConfigAck(DeviceConfigAckMsg),
    PairAck(PairAckMsg),
    ProviderProfiles(ProviderProfilesMsg),
    ProviderProfileAck(ProviderProfileAckMsg),
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn rt(v: &serde_json::Value) -> BridgeToPhone {
        let msg: BridgeToPhone = serde_json::from_value(v.clone()).unwrap();
        let msg2: BridgeToPhone = serde_json::from_value(serde_json::to_value(&msg).unwrap()).unwrap();
        assert_eq!(msg, msg2, "semantic round-trip");
        msg
    }

    fn session() -> serde_json::Value {
        json!({"id":"s","slug":"sl","cwd":"/w","lastActivity":"t","lineCount":1,"title":null,"project":"p","state":"running","seqHigh":10})
    }
    fn entry() -> serde_json::Value {
        json!({"entryType":"text","content":"hello","timestamp":"t"})
    }

    #[test]
    fn heartbeat_minimal_and_full() {
        let m = rt(&json!({"type":"sessions","machine":"m","sessions":[],"protocolVersion":10}));
        match m {
            BridgeToPhone::Sessions(s) => {
                assert_eq!(s.protocol_version, 10);
                assert_eq!(s.host, None);
                assert_eq!(s.removed_sessions, None);
            }
            _ => panic!(),
        }
        let m = rt(&json!({
            "type":"sessions","machine":"m","host":"service","sessions":[session()],
            "authStatus":{"hasAnthropicKey":true,"hasGithubPat":false,"hasEnvKey":false},
            "protocolVersion":10,"capabilities":["sync/1","diff"],
            "folders":["a","b"],"roots":["/w"],"removedSessions":["old"],"machineOffline":true
        }));
        match m {
            BridgeToPhone::Sessions(s) => {
                assert_eq!(s.host, Some(BridgeHostKind::Service));
                assert_eq!(s.removed_sessions.as_deref(), Some(&["old".to_string()][..]));
                assert_eq!(s.machine_offline, Some(true));
            }
            _ => panic!(),
        }
    }

    #[test]
    fn every_variant_decodes_from_a_representative_fixture() {
        rt(&json!({"type":"output","sessionId":"s","seq":7,"entry":entry()}));
        rt(&json!({"type":"input-ack","sessionId":"s","inputId":"i"}));
        rt(&json!({"type":"sync-begin","sessionId":"s","syncId":"y","seqHigh":100,"ranges":[[1,50]]}));
        rt(&json!({"type":"sync-chunk","sessionId":"s","syncId":"y","range":[1,2],"entries":[{"seq":1,"entry":entry()},{"seq":2,"entry":entry()}]}));
        rt(&json!({"type":"sync-end","sessionId":"s","syncId":"y","deliveredRanges":[[1,50]]}));
        rt(&json!({"type":"session-pending","pendingId":"p","machine":"m","createdAt":"t"}));
        rt(&json!({"type":"session-ready","pendingId":"p","session":session()}));
        rt(&json!({"type":"session-failed","pendingId":"p","reason":"boom"}));
        rt(&json!({"type":"input-failed","sessionId":"s","reason":"no-session","inputId":"i"}));
        rt(&json!({"type":"close-session-ack","sessionId":"s","success":true}));
        rt(&json!({"type":"session-replaced","oldSessionId":"o","newSession":session()}));
        rt(&json!({"type":"mode-confirmed","sessionId":"s","mode":"plan"}));
        rt(&json!({"type":"effort-confirmed","sessionId":"s","level":"max"}));
        rt(&json!({"type":"model-confirmed","sessionId":"s","model":"m"}));
        rt(&json!({"type":"folder-ack","requestId":"r","success":true,"path":"a/b"}));
        rt(&json!({"type":"usage","sessionId":"s","usage":{"available":true,"subscriptionType":null,"fetchedAt":"t"}}));
        rt(&json!({"type":"gsd-state","sessionId":"s","gsd":{
            "installed":true,"available":true,"hasGit":true,"situation":"x","summary":"y",
            "milestone":null,"currentPhase":null,"totalPhases":null,"percent":0.0,
            "phases":[],"actions":[],"recommended":null,"paused":false,"blockers":[],
            "verifyFailed":false,"execution":null
        }}));
        rt(&json!({"type":"models","models":[{"id":"m1","label":"M1"},{"id":"m2"}],"defaultModel":"m1"}));
        rt(&json!({"type":"models","models":[],"error":"sdk offline"}));
        rt(&json!({"type":"credentials-ack","machine":"m","success":true,"hasAnthropicKey":true,"hasGithubPat":false,"keyValid":true}));
        rt(&json!({"type":"device-config-ack","success":false,"error":"unreachable"}));
        rt(&json!({"type":"pair-ack","machine":"m","ok":false,"reason":"bad-token","relays":["wss://r"],"host":"cli"}));
        rt(&json!({"type":"provider-profiles","machine":"m","profiles":[{"id":"p","label":"L","baseUrl":"https://x","models":[{"id":"m"}],"hasToken":true}]}));
        rt(&json!({"type":"provider-profile-ack","machine":"m","profileId":"p","success":true,"tokenValid":false}));
    }

    #[test]
    fn unknown_type_is_a_decode_error() {
        assert!(serde_json::from_str::<BridgeToPhone>(r#"{"type":"warp","sessionId":"s"}"#).is_err());
    }

    #[test]
    fn diff_output_entry_survives_the_full_message() {
        let m = rt(&json!({"type":"output","sessionId":"s","seq":1,"entry":{
            "entryType":"diff","content":"+a","timestamp":"t",
            "diff":{"path":"f","lines":[{"type":"add","text":"a"}],"truncated":true}
        }}));
        match m {
            BridgeToPhone::Output(o) => {
                assert_eq!(o.entry.entry_type, crate::wire::common::OutputEntryType::Diff);
                assert_eq!(o.entry.diff.unwrap().truncated, Some(true));
            }
            _ => panic!(),
        }
    }
}
